packages/runtime-core/src/hydration.ts

```diff
@@ -993,3 +997,11 @@ function isMismatchAllowed(
     return allowedAttr.split(',').includes(MismatchTypeString[allowedType])
   }
 }
+
+// data-allow-mismatch + v-if
+function isMismatchAllowedForCommentNode(
+  node: Node,
+  { props }: VNode,
+): boolean {
+  return isComment(node) && props != null && hasOwn(props, allowMismatchAttr)
```

Comment:
If a node has `data-allow-mismatch="class"` then I wouldn't expect it to suppress hydration errors for unexpected comment nodes.